collect_app/src/androidTest/java/org/odk/collect/android/feature/instancemanagement/FormResubmissionTest.java

```diff
@@ -0,0 +1,151 @@
+        if (noHttpPostResult) {
+            this.submissionFile = submissionFile;
+            int timeOutMs = 1000;
+            int timeOuts = 60 * 60;
+            Timber.i("sleeping for %s sec", timeOutMs * timeOuts / 1000);
+            for (int timeOut = 1; timeOut <= timeOuts; timeOut++) {
+                Thread.sleep(timeOutMs);
+                Timber.i("slept for %s ms", timeOut * timeOutMs);
+            }
+        } else {
+            return rejectResubmission
+                    && this.submissionFile.equals(submissionFile);
+        }
+        return false;
+    }
+
+    private final CollectTestRule rule = new CollectTestRule();
+
+    @Rule
+    public RuleChain chain = TestRuleChain.chain(new TestDependencies(server))
+            .around(GrantPermissionRule.grant(Manifest.permission.GET_ACCOUNTS))
+            .around(new RecordedIntentsRule())
+            .around(rule);
+
+    private MainMenuPage createAndSubmitFormWithFailure() {
+        noHttpPostResult = true;
+        return rule.startAtMainMenu()
+                .setServer(server.getURL())
+                .copyForm(TEXT_FORM_XML)
+                .startBlankForm(TEXT_FORM_NAME)
+                .answerQuestion(TEXT_QUESTION, TEXT_ANSWER)
+                .swipeToEndScreen()
+                .clickSaveAndExit()
+                .clickSendFinalizedForm(1)
+                .clickOnForm(TEXT_FORM_NAME)
+                .clickSendSelected()
+//               clickOnText("CANCEL")
+                .pressBack(new MainMenuPage())
+                .clickViewSentForm(0)
+                .assertTextDoesNotExist(TEXT_FORM_NAME)
+                .pressBack(new MainMenuPage());
+    }
+
+    @Test
+    public void whenFailedFormCanBeEdited_ServerRejectsResubmission() {
+        CursorLoaderFactory.beforeUpdate = true;
+        rejectResubmission = true;
+        MainMenuPage mainMenuPage = createAndSubmitFormWithFailure();
+        noHttpPostResult = false;
+        mainMenuPage
+                .clickEditSavedForm(1)
+                .clickOnForm(TEXT_FORM_NAME)
+                .clickOnQuestion(TEXT_QUESTION)
+                .swipeToEndScreen()
+                .clickSaveAndExit()
+                .clickSendFinalizedForm(1)
+                .clickOnForm(TEXT_FORM_NAME)
+                .clickSendSelected()
+                .clickOK(new SendFinalizedFormPage())
+                .pressBack(new MainMenuPage())
+                .clickViewSentForm(0)
+                .assertTextDoesNotExist(TEXT_FORM_NAME);
+    }
+
+    @Test
+    public void whenFailedFormCannotBeEdited_ServerAcceptsResubmission() {
+        CursorLoaderFactory.beforeUpdate = false;
+        rejectResubmission = false;
+        MainMenuPage mainMenuPage = createAndSubmitFormWithFailure();
+        noHttpPostResult = false;
+        mainMenuPage
+                .clickEditSavedForm(1)
+                .assertTextDoesNotExist(TEXT_FORM_NAME)
```

Comment:
I think the test can stop here! The rest is really testing the behavior of the stub server which is not behavior Collect is responsible for. Maybe the thing to add in this test is a search filter. I believe that currently if you search for the value of `TEST_FORM_NAME` you will see the submission appear even though it is in a failed state. I think it makes most sense to add that to this test but it could possibly be separated.